Components/FlightControl/Inc/RocketSM.hpp

```diff
@@ -0,0 +1,295 @@
+ *
+ *        State specific commands, must be all in-order to avoid duplicate command IDs
+ */
+enum RocketControlCommands
+{
+    //-- PRE-IGNITION and RECOVERY --
+    RSC_ABORT,       // Transition to ABORT state - available from all states except for IGNITION/LAUNCH/BURN
+    RSC_OPEN_VENT,   // Open the vent valve
+    RSC_CLOSE_VENT,  // Close the vent valve
+    RSC_OPEN_DRAIN,  // Open the drain valve
+    RSC_CLOSE_DRAIN, // Close the drain valve
+    RSC_MEV_CLOSE,   // Forces MEV to close - ONLY supported in states where it is safe to close the MEV
+
+    //-- PRELAUNCH --
+    RSC_BEGIN_FILL, // Transition to the FILL state
+
+    //-- FILL --
+    RSC_ARM_CONFIRM_1,   // Enable first ARM confirmation flag
+    RSC_ARM_CONFIRM_2,   // Enable second ARM confirmation flag
+    RSC_ARM_ACTION,      // Transition to the ARM state
+
+    //-- ARM/IGNITION/LAUNCH/BURN --
+    RSC_MANUAL_OVERRIDE_ENABLE, // For any states with locked out capability. If this flag is set using a command the NEXT command will be allowed to directly control valves/vents
+
+    //-- ARM --
+    RSC_POWER_TRANSITION_ONBOARD,      // Change power source to onboard
+    RSC_POWER_TRANSITION_EXTERNAL,     // Change power source to external power
+    RSC_READY_FOR_IGNITION, // Ready for ignition sequence - Transition to IGNITION state
+
+    //-- IGNITION --
+    RSC_CONFIRM_IGNITION,   // Confirm igniter actuation - Transition to LAUNCH state (MEV OPEN)
+    //* TBD - To ensure we don't get stuck unable to vent, we have override actions allowing a transition to ABORT
+    RSC_CRITICAL_IGNITION_ALLOW_ABORT, // Enable flag allowing us to abort in the ignition state
+    
+
+    //-- LAUNCH/BURN --
+    //* TBD - To ensure we don't get stuck unable to vent, we have override actions for LAUNCH/BURN
+    RSC_CRITICAL_MEV_CLOSE_OVERRIDE_CONFIRM, // Enable confirmation flag for MEV close override
+    RSC_CRITICAL_MEV_CLOSE_OVERRIDE_ACTION,  // Action for MEV override, requires override flag is set
+    RSC_CRITICAL_BURN_ALLOW_ABORT, // Enable flag allowing us to abort in the ignition state
+    RSC_CRITICAL_CLEAR_OVERRIDE_FLAGS,      // Clears any override flags in either IGNITION/LAUNCH/BURN
+
+    //-- ABORT --
+    RSC_TRANSITION_PRELAUNCH, // Confirm transition back into prelaunch state
+
+    //-- GENERAL --
+    RSC_PAUSE_LOGGING,      // Disable logging - this is supported by ALL states (should be handled in parent class)
+    RSC_START_LOGGING,      // Enable logging - this is supported by ALL states (should be handled in parent class)
+
+    //-- TECHNICAL --
+    RSC_NONE   // Invalid command, must be last
+};
+
+/**
+ * @brief Base class for Rocket State Machine
+ */
+class BaseRocketState
+{
+public:
+    virtual RocketState HandleCommand(Command& cm) = 0; //Handle a command based on the current state
+    virtual RocketState OnEnter() = 0;  //Returns the state we're entering
```

Comment:
really excited about this! I didn't like the idea of having an array of states, but I went through the FSM example with ETL and it looked really clean.

On another note, I'm assuming OnEnter() would be used for broadcasting state changes to threads and starting timers. I also feel like OnExit() might not be used, but its not bad to have for now

Reply:
Yes, most functionality of a given state (not related to commands) will be implemented inside EnterState.

OnExit is used for things like clearing flags or general cleanup.

For example, in the Fill:: state we use two arm confirmation flags, we clear those once we exit that state.